### ISSUE TITLE
FlutterInappPurchase.connectionUpdated not working

### ISSUE THREAD

**Issue by ThangNguyen91** [dooboolab/flutter_inapp_purchase#272]
### Version of flutter_inapp_purchase
flutter_inapp_purchase: ^3.0.1
### Platforms you faced the error (IOS or Android or both?)
both Android and IOS
### Expected behavior
FlutterInappPurchase.connectionUpdated working
### Actual behavior
FlutterInappPurchase.connectionUpdated not working when call await FlutterInappPurchase.instance.initConnection
### Tested environment (Emulator? Real Device?)
Iphone 7plus (Real Device), Samsung j7 (Real Device)
### Steps to reproduce the behavior
Step 1:
var result = await FlutterInappPurchase.instance.initConnection;
Step 2:
_conectionSubscription =
        FlutterInappPurchase.connectionUpdated.listen((connected) {
          _connectedResult = connected;
      print('FlutterInappPurchase connected $connected');
    });

**Comment by arisbartee** [comment]
Seems to be working now.

**Comment by deakjahn** [comment]
I have something even more sinister right now. The `initConnection` handler calls back

    channel.invokeMethod("connection-updated", item.toString());

And it seems that this not only doesn't arrive to the Dart side but it basically hangs the process and it never actually goes on to the next line.

**Comment by hyochan** [comment]
> line

Oh may I know which version you are using?

**Comment by deakjahn** [comment]
Always the last one, 5.1.0.

Now I made a local copy and I keep investigating but practically, the whole channel communication seems to have problems. Strange.

**Comment by deakjahn** [comment]
Basically, this is my local code now:

    Future<String?> initConnection() async {
      if (_platform.isAndroid) {
        await _setPurchaseListener();
        print("before call");
        final String? result = await _channel.invokeMethod('initConnection');
        print("never reaches here");
        return result;
      } else ...
    }

**Comment by hyochan** [comment]
@deakjahn If I remember correctly, there were lots of question asked for module not being updated. Could you kindly remove the cache and ensure that you have the fresh `FlutterInappPurchase` module?

**Comment by deakjahn** [comment]
I have no cache issues for sure. Right know I'm working on a local copy of the module and all `print()` and `Log.d()` printouts I put into the source are immedately picked up on the next compile, so there's no way for anything stale to sneak in. :-)

**Comment by deakjahn** [comment]
On an unrelated note, because I checked out the plugin registration as well, I'd suggest to include the same checks into `registerWith` as well:

  public static void registerWith(Registrar registrar) {
    isAndroid = isPackageInstalled(registrar.context(), "com.android.vending");
    isAmazon = isPackageInstalled(registrar.context(), "com.amazon.venezia");

    // In the case of an amazon device which has been side loaded with the Google Play store,
    // we should use the store the app was installed from.
    if (isAmazon && isAndroid) {
      if (isAppInstalledFrom(registrar.context(), "amazon")) {
        isAndroid = false;
      } else {
        isAmazon = false;
      }
    }

or even better, put the check into a separate function and just call from both places.

**Comment by deakjahn** [comment]
Well, nothing to do with my trouble, but here it is if you're interested, I did it during my attempts. :-)

The common function:

    private void onAttached(Context context, BinaryMessenger messenger) {
      isAndroid = isPackageInstalled(context, "com.android.vending");
      isAmazon = isPackageInstalled(context, "com.amazon.venezia");
  
      // In the case of an amazon device which has been side loaded with the Google Play store,
      // we should use the store the app was installed from.
      if (isAmazon && isAndroid) {
        if (isAppInstalledFrom(context, "amazon")) {
          isAndroid = false;
        } else {
          isAmazon = false;
        }
      }
  
      if (isAndroid) {
        androidInappPurchasePlugin = new AndroidInappPurchasePlugin();
        androidInappPurchasePlugin.setContext(context);
        setupMethodChannel(messenger, androidInappPurchasePlugin);
      } else if (isAmazon) {
        amazonInappPurchasePlugin = new AmazonInappPurchasePlugin();
        amazonInappPurchasePlugin.setContext(context);
        setupMethodChannel(messenger, amazonInappPurchasePlugin);
      }
    }

and both just call this:

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
      onAttached(binding.getApplicationContext(), binding.getBinaryMessenger());
    }

    public static void registerWith(Registrar registrar) {
      FlutterInappPurchasePlugin plugin = new FlutterInappPurchasePlugin();
      plugin.onAttached(registrar.context(), registrar.messenger());
    }

**Comment by deakjahn** [comment]
@hyochan OK. I found it, finally. You won't like it that much but you'll have to change it. :-) First I'll describe it, then I'll try to look whether I can provide you with a PR. My current code is rather different from your current one, I tweaked it quite a lot but I'll try to incorporate what's needed in your original code.

The underlying problem is relatively new:

- https://github.com/flutter/flutter/issues/34993
- https://flutter.dev/docs/development/platform-integration/platform-channels#channels-and-platform-threading

Plugins now have to handle threading and make sure that the report the result on the UI thread even if their actual platform code runs on a different one. One good solution is the `MethodResultWrapper` that the geocoder plugin also applies, this uses a `Runnable` to run the callbacks (both you calling back to the Dart side and reporting success or failure in `result`) on the UI thread.
